Feature: a widget for setting Reader settings

psobolewskiPhD opened the pull request:
This PR adds a second `contribution` to the plugin: a dock widget.
The dock widget allows for setting some parameters that are used by the reader.
The aim is to address #29 but could be extend to #34 
This is a WIP, at present it looks like:
<img width="1202" alt="image" src="https://user-images.githubusercontent.com/76622105/198850616-bbea75ac-82d4-4497-92c8-f9d9abd82162.png">

Things to consider:
1. adding a checkbox to just override the thresholds and load out of memory. This could be accomplished by adding `IN_MEMORY` to `core.py` and setting that as the default for `in_memory` for `reader_function`
2. to address #34 should add `AUTOLOAD_FIRST_SCENE` to `core.py` and modify `_get_scenes` to load the first scene based on this boolean setting.
3. For the `Delimiter`, should the white spaces be set by the user or appended to the string provided by the user?
~~4. I'd like the widget to close itself when the user confirms/sets the settings, but this is still WIP—not sure how to accomplish it.~~ *DONE*
5. Persistence of settings: should they be autoloaded in some fashion when loading the reader?
6. Do we need a button to reset back to the original defaults?

Feedback very welcome!
CC: @lambda-science

Czaki commented on napari_aicsimageio/_settings_widget.py:
```diff
@@ -0,0 +1,17 @@
+
+    napari_aicsimageio.core.SCENE_LABEL_DELIMITER = delimiter
+    napari_aicsimageio.core.IN_MEM_THRESHOLD_SIZE_BYTES = in_mem_size * 1e9
+    napari_aicsimageio.core.IN_MEM_THRESHOLD_PERCENT = frac_mem_size / 100
```
```suggestion
from magicgui.widgets import Container, TextEdit, FloatSpinBox, SpinBox
from napari_aicsimageio import core as aicore


class SetSetings(Container):
    def __init__(self, napari_viewer: "napari.Viewer"):
        self.viewer = napari_viewer
        self.delimiter = TextEdit(label="Delimiter in scene labels:", value = aicore.SCENE_LABEL_DELIMITER) 
        self.in_mem_size = FloatSpinBox(label="Threshold for out-of-memory loading (GB)", value=aicore.IN_MEM_THRESHOLD_SIZE_BYTES/1e9)
        self.frac_mem_size = SpinBox(label="Threshold for out-of-memory loading (% free memory):", value=aicore.IN_MEM_THRESHOLD_PERCENT*100)
        self.call_btn = PushButton("Set Reader Settings")
        self.call_btn.clicked.connect(self.save)
        super().__init__(widgets=[self.delimiter, self.in_mem_size, self.frac_mem_size, self.call_btn])
    
    def save(self):
        aicore.SCENE_LABEL_DELIMITER = self.delimiter.value
        aicore.IN_MEM_THRESHOLD_SIZE_BYTES = self.in_mem_size.value * 1e9
        aicore.IN_MEM_THRESHOLD_PERCENT = self.frac_mem_size.value / 100
        self.viewer.window.remove_dock_widget(self.native)
```

Czaki commented on napari_aicsimageio/napari.yaml:
```diff
@@ -5,6 +5,13 @@ contributions:
     python_name: napari_aicsimageio.core:get_reader
+  - id: napari-aicsimageio.settings_widget
+    title: Open AICSImageIO Reader Settings
+    python_name: napari_aicsimageio._settings_widget:set_settings
```
```suggestion
    python_name: napari_aicsimageio._settings_widget:SetSettings
```

psobolewskiPhD commented:
I think the lint error requires updating flake8 to >5...

psobolewskiPhD commented:
I've updated the widget based on @Czaki tips above, but sticking with magic_factory.
The default values are now read from `core.py` rather than hard-coded.
The widget values will now persist between instances of the widget and napari sessions.
HOWEVER, to apply the settings you still need to open the widget and run it.

In theory, could use the same approach as magicgui to look for the cached settings and read them, but not on widget load  but instead on-load of the reader. This would make it more intuitive that the settings are set and forget, unless you want to change them.
Hacky.

Czaki commented:
> The default values are now read from `core.py` rather than hard-coded.
> The widget values will now persist between instances of the widget and napari sessions.

Yes, but on open of widget they will not be consistent with values from`napari_aicsimageio.core` so that may introduce confusion of users who check values from the widget, they look correct and do not notice that they need to be fixed. But the result of using readers will be incorrect.

psobolewskiPhD commented:
> > The default values are now read from `core.py` rather than hard-coded.
> > The widget values will now persist between instances of the widget and napari sessions.
> 
> Yes, but on open of widget they will not be consistent with values from`napari_aicsimageio.core` so that may introduce confusion of users who check values from the widget, they look correct and do not notice that they need to be fixed. But the result of using readers will be incorrect.

Not sure I follow. You mean if the user opens the widget but then doesn't click the button?

Czaki commented:
> Not sure I follow. You mean if the user opens the widget but then doesn't click the button?

Yes

psobolewskiPhD commented:
Right, I agree it could be confusing. 
The only way to change that behavior is to look for cached magicgui file and read & apply on-load.

Czaki commented:
> Right, I agree it could be confusing. 
> The only way to change that behavior is to look for cached magicgui file and read & apply on-load.

Or stop using magic factory for something that is not a function as it is something that mutates global state.

psobolewskiPhD commented:
But mutating the global state is the goal of this PR?
It's not perfect, but it's not terrible?
If the user wants to change the memory behavior of the reader, they use the widget and apply their settings.
If they don't they get defaults.
Next session, if they want to come back to their settings, they're stored in the widget, but still need to be applied. If not, they have the defaults.
This isn't horrible I don't think.

psobolewskiPhD commented:
I've added a header to inform that the button needs to be pressed per session and changed the button text to use the verb `Apply` which I think is clearer.
<img width="478" alt="image" src="https://user-images.githubusercontent.com/76622105/198903247-8904a039-2c3d-4ad1-b9ef-60d3ce646cbb.png">

evamaxfield commented on napari_aicsimageio/_settings_widget.py:
```diff
@@ -0,0 +1,33 @@
+    frac_mem_size={
+        "label": "Threshold for out-of-memory loading (<small>% free memory</small>):"
+    },
+    call_button="Apply Reader Settings and Close Widget",
```
Does this _need_ to close the widget? I am tempted to keep it open the whole time :woman_shrugging:

psobolewskiPhD replied:
I have a small 13" screen, so Apply and close -> drag and drop -> scene selector was more comfortable.
Maybe I can add a checkbox for `Close on Apply`?

evamaxfield replied:
Can there be two buttons next to each other? [Apply] <-- space --> [Apply and Close]

evamaxfield replied:
eh. another checkbox is fine too

psobolewskiPhD replied:
I'll take a look 😊

psobolewskiPhD replied:
I don't like 2 buttons though. I think `Apply` and then people can close the widget manually is probably the most rational option.

evamaxfield reviewed (commented):
Sorry for the delay. I just want to say this is so simple I love it!! Thanks for this. Few comments

evamaxfield commented:
> Or stop using magic factory for something that is not a function as it is something that mutates global state.

@Czaki I personally think this is one of those: "in a perfect world where we have a lot of time on our hands", sure I agree with you, but the fact that this is so simple and works I think is pretty great. -- I personally haven't had the time to work on this lib or base aicsimageio in a long time so I will take what I can get.

I guess we could write our own state manager but that sounds like overhead. Pinging @tlambert03 if you know of any easy tricks to solve.

My only other comment on this unless we are presented with a better state management solution is that I think the state widget should be kept open after apply. And the button text should just be "Apply".

Czaki commented:
> I guess we could write our own state manager but that sounds like overhead. Pinging @tlambert03 if you know of any easy tricks to solve.

it depends on what you want to have. I could provide some solution (simple use of `appdirs` for the correct location and json for serialize/deserialize). 

The main decision is how to handle different versions of the library. Did you expect that state could strongly change in the new releases?

evamaxfield commented:
> it depends on what you want to have. I could provide some solution (simple use of appdirs for the correct location and json for serialize/deserialize).
> 
> The main decision is how to handle different versions of the library. Did you expect that state could strongly change in the new releases?

These are good questions. And utilization of appdirs would be nice.

After the opening of #67, maybe it makes sense to take a step back and form a single widget that manages all of this state. Not two widgets. Additionally to handle: "how we can grow the options that are managed by state".

I know I will want to add more settings in the future. Particularly: "Chunk Dimensions: {user string}" (Defaults to ZYX).

Sorry @psobolewskiPhD let me think a bit more on this.

psobolewskiPhD commented:
> > it depends on what you want to have. I could provide some solution (simple use of appdirs for the correct location and json for serialize/deserialize).
> > The main decision is how to handle different versions of the library. Did you expect that state could strongly change in the new releases?
> 
> These are good questions. And utilization of appdirs would be nice.
> 
> After the opening of #67, maybe it makes sense to take a step back and form a single widget that manages all of this state. Not two widgets. Additionally to handle: "how we can grow the options that are managed by state".
> 
> I know I will want to add more settings in the future. Particularly: "Chunk Dimensions: {user string}" (Defaults to ZYX).
> 
> Sorry @psobolewskiPhD let me think a bit more on this.

No problem!
I think it's good to have a vision of how this should work.
I still like the idea of setting preferences/settings independently of loading scenes, etc.
Whether that's a widget or a preference pane (in napari) or what, that's what I'd vote for.

But the file specific stuff, that the Scene Management widget enables (unpack channels, clear) I like the way it is, where you can do it file by file. I just have a small monitor to #66 bites!

evamaxfield commented:
Made a comment here: https://forum.image.sc/t/napari-performance-on-a-tb-czi-image/84587/12

lets ignore any sort of persistance for now and at least get this out?
